pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
+	c, err := cycleState.Read(preFilterStateKey)
+	if err != nil {
+		// preFilterState doesn't exist, likely PreFilter wasn't invoked.
+		return nil, fmt.Errorf("error reading %q from cycleState: %w", preFilterStateKey, err)
+	}
+
+	s, ok := c.(*preFilterState)
+	if !ok {
+		return nil, fmt.Errorf("%+v convert to volumerestrictions.state error", c)
+	}
+	return s, nil
+}
+
+// calPreFilterState computes preFilterState describing which PVCs use ReadWriteOncePod
+// and which pods in the cluster are in conflict.
+func (pl *VolumeRestrictions) calPreFilterState(ctx context.Context, pod *v1.Pod) (*preFilterState, error) {
+	pvcs, err := pl.readWriteOncePodPVCsForPod(ctx, pod)
+	if err != nil {
+		return nil, err
 	}
-	return nil, framework.NewStatus(framework.Success)
+
+	pvcNames := sets.NewString()
+	for _, pvc := range pvcs {
+		pvcNames.Insert(pvc.Name)
+	}
+
+	pods, err := pl.podLister.Pods(pod.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	conflictingPods := make(map[string]*v1.Pod)
+	for _, p := range pods {
+		for _, volume := range p.Spec.Volumes {
+			// The current pod cannot be in conflict with itself.
+			if p.Name == pod.Name {
+				continue
+			}
+			if volume.PersistentVolumeClaim == nil {
+				continue
+			}
+
+			pvcName := volume.PersistentVolumeClaim.ClaimName
+			if pvcNames.Has(pvcName) {
+				conflictingPods[p.Name] = p
+			}
+		}
+	}
+	return &preFilterState{
+		readWriteOncePodPVCs: pvcNames,
+		conflictingPods:      conflictingPods,
+	}, nil
 }
 
-// isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
-// This access mode restricts volume access to a single pod on a single node. Since only a single pod can
-// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as UnschedulableAndUnresolvable.
-// TODO(#103132): Mark pod as Unschedulable and add preemption logic.
-func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(ctx context.Context, pod *v1.Pod) *framework.Status {
+func (pl *VolumeRestrictions) readWriteOncePodPVCsForPod(ctx context.Context, pod *v1.Pod) ([]*v1.PersistentVolumeClaim, error) {
+	pvcs := []*v1.PersistentVolumeClaim{}
 	for _, volume := range pod.Spec.Volumes {
 		if volume.PersistentVolumeClaim == nil {
 			continue
 		}
 
 		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(volume.PersistentVolumeClaim.ClaimName)
 		if err != nil {
-			if apierrors.IsNotFound(err) {
-				return framework.NewStatus(framework.UnschedulableAndUnresolvable, err.Error())
-			}
-			return framework.AsStatus(err)
+			return nil, err
 		}
 
 		if !v1helper.ContainsAccessMode(pvc.Spec.AccessModes, v1.ReadWriteOncePod) {
 			continue
 		}
+		pvcs = append(pvcs, pvc)
+	}
+	return pvcs, nil
+}
+
+// Checks if scheduling the pod onto this node would cause any conflicts with
+// existing volumes.
+func satisfyExistingVolumes(pod *v1.Pod, nodeInfo *framework.NodeInfo) bool {
```

Comment:
Can't think of a better name, but do you have other alternatives?

Reply:
I changed it to `satisfyVolumeConflicts` but I don't really feel great about that either